README.md
## Scrutinizer Datasource - a datasource for Plixer's Scrutinizer 

More documentation about this Datasource can be found here [Docs](https://www.plixer.com).






CHANGELOG.md
# Version 2 (11/6/2020)

## Features

+ Expanded adhoc filter capability to include the ability to Select Granularity, Resolve DNS, and Hide Others. 

## Bug Fixes 

+ Fixed issue where Adhoc Filters for Device Groups were not being passed correctly resulting in incorrect data being returned. 



spec/datasource_spec.js
// import {Datasource} from "../module";
// import Q from "q";

// describe('GenericDatasource', function() {
//     var ctx = {};

//     beforeEach(function() {
//         ctx.$q = Q;
//         ctx.backendSrv = {};
//         ctx.templateSrv = {};
//         ctx.ds = new Datasource({}, ctx.$q, ctx.backendSrv, ctx.templateSrv);
//     });

//     it('should return an empty array when no targets are set', function(done) {
//         ctx.ds.query({targets: []}).then(function(result) {
//             expect(result.data).to.have.length(0);
//             done();
//         });
//     });

//     it('should return the server results when a target is set', function(done) {
//         ctx.backendSrv.datasourceRequest = function(request) {
//             return ctx.$q.when({
//                 _request: request,
//                 data: [
//                     {
//                         target: 'X',
//                         datapoints: [1, 2, 3]
//                     }
//                 ]
//             });
//         };

//         ctx.templateSrv.replace = function(data) {
//             return data;
//         }

//         ctx.ds.query({targets: ['hits']}).then(function(result) {
//             expect(result._request.data.targets).to.have.length(1);

//             var series = result.data[0];
//             expect(series.target).to.equal('X');
//             expect(series.datapoints).to.have.length(3);
//             done();
//         });
//     });

//     it ('should return the metric results when a target is null', function(done) {
//         ctx.backendSrv.datasourceRequest = function(request) {
//             return ctx.$q.when({
//                 _request: request,
//                 data: [
//                     "metric_0",
//                     "metric_1",
//                     "metric_2",
//                 ]
//             });
//         };

//         ctx.templateSrv.replace = function(data) {
//             return data;
//         }

//         ctx.ds.metricFindQuery({target: null}).then(function(result) {
//             expect(result).to.have.length(3);
//             expect(result[0].text).to.equal('metric_0');
//             expect(result[0].value).to.equal('metric_0');
//             expect(result[1].text).to.equal('metric_1');
//             expect(result[1].value).to.equal('metric_1');
//             expect(result[2].text).to.equal('metric_2');
//             expect(result[2].value).to.equal('metric_2');
//             done();
//         });
//     });

//     it ('should return the metric target results when a target is set', function(done) {
//         ctx.backendSrv.datasourceRequest = function(request) {
//             var target = request.data.target;
//             var result = [target + "_0", target + "_1", target + "_2"];

//             return ctx.$q.when({
//                 _request: request,
//                 data: result
//             });
//         };

//         ctx.templateSrv.replace = function(data) {
//             return data;
//         }

//         ctx.ds.metricFindQuery('search').then(function(result) {
//             expect(result).to.have.length(3);
//             expect(result[0].text).to.equal('search_0');
//             expect(result[0].value).to.equal('search_0');
//             expect(result[1].text).to.equal('search_1');
//             expect(result[1].value).to.equal('search_1');
//             expect(result[2].text).to.equal('search_2');
//             expect(result[2].value).to.equal('search_2');
//             done();
//         });
//     });

//     it ('should return the metric results when the target is an empty string', function(done) {
//         ctx.backendSrv.datasourceRequest = function(request) {
//             return ctx.$q.when({
//                 _request: request,
//                 data: [
//                     "metric_0",
//                     "metric_1",
//                     "metric_2",
//                 ]
//             });
//         };

//         ctx.templateSrv.replace = function(data) {
//             return data;
//         }

//         ctx.ds.metricFindQuery('').then(function(result) {
//             expect(result).to.have.length(3);
//             expect(result[0].text).to.equal('metric_0');
//             expect(result[0].value).to.equal('metric_0');
//             expect(result[1].text).to.equal('metric_1');
//             expect(result[1].value).to.equal('metric_1');
//             expect(result[2].text).to.equal('metric_2');
//             expect(result[2].value).to.equal('metric_2');
//             done();
//         });
//     });

//     it ('should return the metric results when the args are an empty object', function(done) {
//         ctx.backendSrv.datasourceRequest = function(request) {
//             return ctx.$q.when({
//                 _request: request,
//                 data: [
//                     "metric_0",
//                     "metric_1",
//                     "metric_2",
//                 ]
//             });
//         };

//         ctx.templateSrv.replace = function(data) {
//             return data;
//         }

//         ctx.ds.metricFindQuery().then(function(result) {
//             expect(result).to.have.length(3);
//             expect(result[0].text).to.equal('metric_0');
//             expect(result[0].value).to.equal('metric_0');
//             expect(result[1].text).to.equal('metric_1');
//             expect(result[1].value).to.equal('metric_1');
//             expect(result[2].text).to.equal('metric_2');
//             expect(result[2].value).to.equal('metric_2');
//             done();
//         });
//     });

//     it ('should return the metric target results when the args are a string', function(done) {
//         ctx.backendSrv.datasourceRequest = function(request) {
//             var target = request.data.target;
//             var result = [target + "_0", target + "_1", target + "_2"];

//             return ctx.$q.when({
//                 _request: request,
//                 data: result
//             });
//         };

//         ctx.templateSrv.replace = function(data) {
//             return data;
//         }

//         ctx.ds.metricFindQuery('search').then(function(result) {
//             expect(result).to.have.length(3);
//             expect(result[0].text).to.equal('search_0');
//             expect(result[0].value).to.equal('search_0');
//             expect(result[1].text).to.equal('search_1');
//             expect(result[1].value).to.equal('search_1');
//             expect(result[2].text).to.equal('search_2');
//             expect(result[2].value).to.equal('search_2');
//             done();
//         });
//     });

//     it ('should return data as text and as value', function(done) {
//         var result = ctx.ds.mapToTextValue({data: ["zero", "one", "two"]});

//         expect(result).to.have.length(3);
//         expect(result[0].text).to.equal('zero');
//         expect(result[0].value).to.equal('zero');
//         expect(result[1].text).to.equal('one');
//         expect(result[1].value).to.equal('one');
//         expect(result[2].text).to.equal('two');
//         expect(result[2].value).to.equal('two');
//         done();
//     });

//     it ('should return text as text and value as value', function(done) {
//         var data = [
//             {text: "zero", value: "value_0"},
//             {text: "one", value: "value_1"},
//             {text: "two", value: "value_2"},
//         ];

//         var result = ctx.ds.mapToTextValue({data: data});

//         expect(result).to.have.length(3);
//         expect(result[0].text).to.equal('zero');
//         expect(result[0].value).to.equal('value_0');
//         expect(result[1].text).to.equal('one');
//         expect(result[1].value).to.equal('value_1');
//         expect(result[2].text).to.equal('two');
//         expect(result[2].value).to.equal('value_2');
//         done();
//     });

//     it ('should return data as text and index as value', function(done) {
//         var data = [
//             {a: "zero", b: "value_0"},
//             {a: "one", b: "value_1"},
//             {a: "two", b: "value_2"},
//         ];

//         var result = ctx.ds.mapToTextValue({data: data});

//         expect(result).to.have.length(3);
//         expect(result[0].text).to.equal(data[0]);
//         expect(result[0].value).to.equal(0);
//         expect(result[1].text).to.equal(data[1]);
//         expect(result[1].value).to.equal(1);
//         expect(result[2].text).to.equal(data[2]);
//         expect(result[2].value).to.equal(2);
//         done();
//     });

//     it('should support tag keys', function(done) {
//         var data =  [{'type': 'string', 'text': 'One', 'key': 'one'}, {'type': 'string', 'text': 'two', 'key': 'Two'}];

//         ctx.backendSrv.datasourceRequest = function(request) {
//             return ctx.$q.when({
//                 _request: request,
//                 data: data
//             });
//         };

//         ctx.ds.getTagKeys().then(function(result) {
//             expect(result).to.have.length(2);
//             expect(result[0].type).to.equal(data[0].type);
//             expect(result[0].text).to.equal(data[0].text);
//             expect(result[0].key).to.equal(data[0].key);
//             expect(result[1].type).to.equal(data[1].type);
//             expect(result[1].text).to.equal(data[1].text);
//             expect(result[1].key).to.equal(data[1].key);
//             done();
//         });
//     });

//     it('should support tag values', function(done) {
//         var data =  [{'key': 'eins', 'text': 'Eins!'}, {'key': 'zwei', 'text': 'Zwei'}, {'key': 'drei', 'text': 'Drei!'}];

//         ctx.backendSrv.datasourceRequest = function(request) {
//             return ctx.$q.when({
//                 _request: request,
//                 data: data
//             });
//         };

//         ctx.ds.getTagValues().then(function(result) {
//             expect(result).to.have.length(3);
//             expect(result[0].text).to.equal(data[0].text);
//             expect(result[0].key).to.equal(data[0].key);
//             expect(result[1].text).to.equal(data[1].text);
//             expect(result[1].key).to.equal(data[1].key);
//             expect(result[2].text).to.equal(data[2].text);
//             expect(result[2].key).to.equal(data[2].key);
//             done();
//         });
//     });

// });


Gruntfile.js
module.exports = function(grunt) {

  require('load-grunt-tasks')(grunt);

  grunt.loadNpmTasks('grunt-execute');
  grunt.loadNpmTasks('grunt-contrib-clean');

  grunt.initConfig({

    clean: ["dist"],

    copy: {
      src_to_dist: {
        cwd: 'src',
        expand: true,
        src: ['**/*', '!**/*.js', '!**/*.scss'],
        dest: 'dist'
      },
      pluginDef: {
        expand: true,
        src: ['README.md'],
        dest: 'dist'
      }
    },

    watch: {
      rebuild_all: {
        files: ['src/**/*'],
        tasks: ['default'],
        options: {spawn: false}
      }
    },

    babel: {
      options: {
        sourceMap: true,
        presets:  ['es2015']
      },
      dist: {
        options: {
          plugins: ['transform-es2015-modules-systemjs', 'transform-es2015-for-of', 'transform-object-rest-spread']
        },
        files: [{
          cwd: 'src',
          expand: true,
          src: ['**/*.js'],
          dest: 'dist',
          ext:'.js'
        }]
      },
      distTestNoSystemJs: {
        files: [{
          cwd: 'src',
          expand: true,
          src: ['**/*.js'],
          dest: 'dist/test',
          ext:'.js'
        }]
      },
      distTestsSpecsNoSystemJs: {
        files: [{
          expand: true,
          cwd: 'spec',
          src: ['**/*.js'],
          dest: 'dist/test/spec',
          ext:'.js'
        }]
      }
    },

    mochaTest: {
      test: {
        options: {
          reporter: 'spec'
        },
        src: ['dist/test/spec/test-main.js', 'dist/test/spec/*_spec.js']
      }
    }
  });

  grunt.registerTask('default', ['clean', 'copy:src_to_dist', 'copy:pluginDef', 'babel', 'mochaTest']);
};


src/testing.js
export class scrutinizerJSON {
  constructor() {}

  reportJSON(authToken, reportType, startTime, endTime, ipAddress) {
    let params = {
      rm: "report_api",
      action: "get",
      authToken: authToken,
      rpt_json: JSON.stringify({
        reportTypeLang: reportType,
        reportDirections: {
          selected: "inbound"
        },
        times: {
          dateRange: "Custom",
          start: `${startTime}`,
          end: `${endTime}`,
          clientTimezone: "America/New_York"
        },
        filters: {
          sdfDips_0: `in_${ipAddress}_${ipAddress}_ALL`
        },
        dataGranularity: {
          selected: "auto"
        }
      }),

      data_requested: JSON.stringify({
        inbound: {
          graph: "all",
          table: {
            query_limit: {
              offset: 0,
              max_num_rows: 10
            }
          }
        }
      })
    };
    
    return params;
  }


}

export class scrutinizerRequest {
  constructor() {
 //scrutinizer returns graph data opposite of how grafana wants it. So we flip it here.
    this.rearrangeData = (arr, oldIndex, newIndex) => {
      while (oldIndex < 0) {
        old_index += arr.length;
      }
      while (newIndex < 0) {
        new_index += arr.length;
      }
      if (newIndex >= arr.length) {
        let k = newIndex - arr.length;
  
        while (k-- + 1) {
          arr.push(undefined);
        }
      }
      arr.splice(newIndex, 0, arr.splice(oldIndex, 1)[0]);
      return arr;
    };
  }

  //used to get all the parameted into the request.
  formatParams(params) {
    return (
      "?" +
      Object.keys(params)
        .map(function(key) {
          return key + "=" + encodeURIComponent(params[key]);
        })
        .join("&")
    );
  }


  //grafana wants time in millaseconds. so we multiple by 1000.
  formatData(scrutData){
    let datatoGraph = [];
    let graphingData = JSON.parse(scrutData.responseText);
    let i, j = 0;
    let graphData = graphingData["report"]["graph"]["pie"]["inbound"];
    let tableData = graphingData["report"]["graph"]["timeseries"]["inbound"];
    for (i = 0; i < tableData.length; i++) {
      for (j = 0; j < tableData[i].length; j++) {
        tableData[i][j][0] = tableData[i][j][0] * 1000;

        this.rearrangeData(tableData[i][j], 0, 1);
      }
    }

    for (i = 0; i < graphData.length; i++) {
      datatoGraph.push({
        target: graphData[i]["label"],
        datapoints: tableData[i]
      });
    }

      return datatoGraph
  }

  makeRequest(scrutUrl, scrutParams) {
    let request = new XMLHttpRequest();
    let url = scrutUrl + scrutParams;
    return new Promise((resolve, reject) => {
      request.onreadystatechange = function() {
        // Only run if the request is complete
        if (request.readyState !== 4) return;

        // Process the response
        if (request.status >= 200 && request.status < 300) {
          // If successful
          console.log(request);

          resolve(request);
        } else {
          // If failed
          reject({
            status: request.status,
            statusText: request.statusText
          })
        }
      };

      // Setup our HTTP request
      request.open("GET", url, true);

      // Send the request
      request.send();
    });
  }
};




src/query_ctrl.js
import {QueryCtrl} from 'app/plugins/sdk';
import './css/query-editor.css!'

export class GenericDatasourceQueryCtrl extends QueryCtrl {
  //creates the link between dropdowns and the datasource
  constructor($scope, $injector)  {

    super($scope, $injector);
    this.scope = $scope;
    this.target.target = this.target.target || 'Select Exporter';
    this.target.report = this.target.report || 'Select Report';
    this.target.display = this.target.display || 'Bits / Percent';
    this.target.direction = this.target.direction || 'Select Direction';
    this.target.interface = this.target.interface || 'Select Interface';
    this.target.type = this.target.type || 'timeserie';
    this.target.filters = this.target.filters;
    this.target.dns = this.target.resolveDNS;
    this.target.granularity = this.target.granularity || 'Select Granularity';
    this.target.entity = this.target.entity || 'Select Entity';
    this.target.entityRows = this.target.entityRows || 'Number of Results';
    this.target.hideOthers = this.target.showOthers;
    

  }
  //each drop down gets a function that is called by the datasource. 
  getOptions(query) {
  

    return this.datasource.getExporters(query || '', this.scope);
  }

  getInterfaces(query) {


    return this.datasource.findInterfaces(query || '', this.scope)
  }
  toggleEditorMode() {
    console.log(query)
    this.target.rawQuery = !this.target.rawQuery;
  }

  onChangeInternal() {
    this.panelCtrl.refresh(); // Asks the panel to refresh data.
  }

  getReports() {

    return this.datasource.reportOptions
  }

  getDisplay() {

    return this.datasource.displayOptions
  }
  
  getDirection() {

    return this.datasource.reportDirections
  }

  getEntities() {


    return this.datasource.entityTypes
  }


  getEntitiesRows() {


    return this.datasource.entityRowCount
  }
  getGranularity(){
    return this.datasource.granularityOptions
  }



  applyFilter() {
    return this.datasource.applyFilter(this.scope, this.panelCtrl)
  }

  resolveDNS() {

    
    this.target.dns = !this.target.dns
    this.panelCtrl.refresh()


  }

  hideOthers() {

    this.target.hideOthers = !this.target.hideOthers;


    this.panelCtrl.refresh()



  }





}


GenericDatasourceQueryCtrl.templateUrl = 'partials/query.editor.html';



src/reportTypes.js
//Update this list to add more report types. 
export const reportTypes = [
  {
    text: "Conversations",
    value: "conversations"
  },
  {
    text: "Applications",
    value: "applications"
  },
  {
    text: "Type of Service",
    value: "tos"
  },
  {
    text: "Source Hosts",
    value: "srcHosts"
  },
  {
    text: "Destination Hosts",
    value: "dstHosts"
  },
  {
    text: "Interface Utilization",
    value: "interfaces"
  }
];
//it's best not to use bi-direction, just build two queries in grafana.
export const reportDirection = [
  {
    text: "Inbound",
    value: "inbound"
  },
  {
    text: "Outbound",
    value: "outbound"
  }
];

//currenlty only used for interface reporting. Other reports the percent utilized it calculated in
//the front end, not sent back in the JSON. It's possible to do this later on, but for now defaulting
//to bits for anything that is not interfaces report. 
export const displayOptions = [
  {
    text: "Bits Per Second",
    value: "bits"
  },
  {
    text: "Percent Utilized",
    value: "percent"
  }
];


export const filterTypes = [
  {
    text: "Source IP Filter",
    value:"sourceIp"
  }, 
  {
    text: "Destination IP Filter",
    value:"destIp"
  }, 
  {
    text: "Add Port Filter",
    value:"ports"
  },
  {
    text: "Show Others",
    value:"others"
  },
  {
    text: "Select Granularity",
    value:"granularity"
  },
  {
    text: "Resolve DNS",
    value:"resolve"
  },
]



export const granularityOptions = [
  {
    text: "Auto",
    value: "auto"
  },
  {
    text: "1 Minute",
    value: "1"
  },
  {
    text: "5 Minute",
    value: "5"
  },
  {
    text: "30 Minute",
    value: "30"
  },
  {
    text: "2 Hour",
    value: "120"
  },
  {
    text: "12 Hour",
    value: "720"
  }
];


export const resolveDNS = [
  {
    text: "Yes",
    value: true
  },
  {
    text: "No",
    value: false
  },

]


src/reportData.js
import _ from "lodash";


export class ScrutinizerJSON {
  constructor() {}
  createFilters (scrut, options, reportFilter,query){
    let {authToken} = scrut;
    let {reportType, reportDirection,reportDisplay } = query
    let scrutDisplay;
    if (reportDisplay === "percent") {
      scrutDisplay = { display: "custom_interfacepercent" };
    } else {
      scrutDisplay = { display: "sum_octetdeltacount" };
    }
    return {
      authToken,
      reportType,
      startTime: options["range"]["from"].unix(),
      endTime: options["range"]["to"].unix(),
      reportDirection,
      scrutDisplay,
      scrutFilters : reportFilter

    }

  }
// used for single query
  createParams (scrut, options, query, adhocFilters = null) {


    let selectedGranularity = query.reportGranularity

    if(!Number.isInteger(parseInt(selectedGranularity))) {
      selectedGranularity = 'auto'
    }
    
    let {authToken} = scrut;
    let {reportType, reportDirection,reportInterface, target, reportFilters, reportDisplay} = query
    let startTime = options["range"]["from"].unix()
    let endTime = options["range"]["to"].unix()
    let scrutFilters;
    let exporterInterface;
    let scrutDisplay;

    // if there are adhoc filters for a device group / exporter user it. Otherwise dont. 
    if (adhocFilters !== null) {
      scrutFilters = adhocFilters
    }else {
      if (reportInterface === "allInterfaces") {
        exporterInterface = "_ALL";
      } else {
        exporterInterface = reportInterface;
      }
  
      //  if user wants all devices, then they are defualted to all interfaces
      if (target === "allExporters") {
        scrutFilters = {
          sdfDips_0: `in_GROUP_ALL`
        };
      } else if (target === "deviceGroup") {
        scrutFilters = {
          sdfDips_0: `in_GROUP_${exporterInterface}`
        };
      } else {
        // if user wants a specific device, they can either have ALL interfaces, or a specific interface
        if (exporterInterface === "_ALL") {
          scrutFilters = {
            sdfDips_0: `in_${target}_ALL`
          };
        } else {
          scrutFilters = {
            sdfDips_0: `in_${target}_${target}-${exporterInterface}`
          };
        }
      }


    }



    
    if (reportFilters !== "No Filter") {
      let filterJson = JSON.parse(reportFilters);
      for (var key in filterJson) {
        if (filterJson.hasOwnProperty(key)) {
          if (key != "sdfDips_0") {
            scrutFilters[key] = filterJson[key];
          }
        }
      }
    }
    if (reportDisplay === "percent") {
      scrutDisplay = { display: "custom_interfacepercent" };
    } else {
      scrutDisplay = { display: "sum_octetdeltacount" };
    }



    return {
        authToken,
        reportType,
        startTime,
        endTime,
        reportDirection,
        scrutFilters,
        scrutDisplay,
        selectedGranularity
       
  
      }
    
     


  }

// cused to create filter object for adhoc queries
createAdhocFilters(filterObject) {


  let reportFilters = {};

  //if there are ip addres filters, add them
  if (filterObject.sourceIp.length > 0) {
    filterObject.sourceIp.forEach((element, index) => {
      let filerCount = `sdfIps_${index}`;
      reportFilters[filerCount] = `in_${element}_src`;
    });
  }

  if (filterObject.destIp.length > 0) {
    filterObject.destIp.forEach((element, index) => {
      let filerCount = `sdfIps_${index}`;
      reportFilters[filerCount] = `in_${element}_dst`;
    });
  }

  if (filterObject.ports.length > 0) {
    filterObject.ports.forEach((element, index) => {
      let filerCount = `sdfSdPorts_${index}`;
      reportFilters[filerCount] = `in_${element}_both`;
    });
  }
  //there will always be exporter filters, add them.
  filterObject.exporterDetails.forEach((element, index) => {
    
    let { exporterIp, interfaceId } = element;

    let filterCount = `sdfDips_${index}`;
    if(exporterIp === "GROUP"){
      reportFilters[filterCount] = `in_${exporterIp}_${interfaceId}`
    }else if (exporterIp === "ALL"){
      reportFilters[filterCount] = `in_${exporterIp}_${interfaceId}`
    } else if (interfaceId != "ALL")
    {
      reportFilters[filterCount] = `in_${exporterIp}_${exporterIp}-${interfaceId}`;
    }else if (exporterIp != "ALL" && exporterIp != "GROUP"){
      reportFilters[filterCount] = `in_${exporterIp}_${interfaceId}`
    }
    
  });
 

  return reportFilters;
}


  
  authJson(scrutInfo){
    return {
      url: scrutInfo['url'],
      method: "GET",
      params:{
        rm: "licensing",
        authToken:scrutInfo['authToken']
      }
    }
  }


  exporterJSON(scrutInfo) {
    //params to figure out which exporters are available to pick from.
    return {
      url: scrutInfo['url'],
      method: "GET",
      params: {
        rm: "get_known_objects",
        type: "devices",
        authToken: scrutInfo['authToken']
      }
    };
  };

  findExporter(scrutInfo, exporter) {
    
    return {
      url: scrutInfo["url"],
      method: "GET",
      params: {
        rm: "loadMap",
        action: "search",
        str: exporter,
        authToken: scrutInfo["authToken"],
        defaultGroupOnTop: 1,
        statusTreeEnabled: 1,
        page: 1
      }
    };
  };

  findtimeJSON(scrutInfo,scrutParams,query, filterObject = null) {

    let selectedGranularity = query.reportGranularity
    if (filterObject){
      if (filterObject.granularity){
        selectedGranularity = filterObject.granularity
      }
      
    }

  
    if(!Number.isInteger(parseInt(selectedGranularity))) {
      selectedGranularity = 'auto'
    }
    

    //params to figure out which interval your in based on data you are requesting
    return {
      url:scrutInfo['url'],
      method:'get',
      params:{
        rm: "report_start",
        authToken: scrutInfo['authToken'],
        report_data: JSON.stringify({
          parse: true,
          reportDirections: { selected: `${scrutParams.reportDirection}` },
          reportTypeLang: `${scrutParams.reportType}`,
          times: {
            dateRange: "Custom",
            start: `${scrutParams.startTime}`,
            end: `${scrutParams.endTime}`,
            clientTimezone: "America/New_York"
          },
          filters: scrutParams.scrutFilters,
          dataGranularity: { selected: selectedGranularity },
          oneCollectorRequest: false
        })
      },

    };
  };

  groupJSON(url, authToken) {
    return {
      url,
      method: "GET",
      params: {
        rm: "get_known_objects",
        type: "deviceGroups",
        authToken
      }
    };
  };

  interfaceJSON(scrutInfo, ipAddress) {

    if(ipAddress['key'] ==="Device Group"){
      let groupName = ipAddress['value']
      return {
        url: scrutInfo['url'],
        method:"get",
        

        params: {
          rm: "mappingConfiguration",
          view: "mapping_configuration",
          authToken: scrutInfo["authToken"],
          session_state:{
            "client_time_zone":"America/New_York","order_by":[],
            "search":[
              {
                "column":"name",
                "value":groupName,
                "comparison":"equal",
                "data":
                  {"filterType":"string"},"_key":"name_equal_Cisco"}
                ],
            "query_limit":{
              "offset":0,"max_num_rows":50},"hostDisplayType":"dns"}
          }
        }
    } else {
      let exporterName
      if (ipAddress['value']) {
        exporterName = ipAddress['value']
      } else {
        exporterName = ipAddress
      }
      
   
      return {
        url: scrutInfo["url"],
        method: "get",
        params: {
          rm: "status",
          action: "get",
          view: "topInterfaces",
          authToken: scrutInfo["authToken"],
          session_state: JSON.stringify({
            client_time_zone: "America/New_York",
            order_by: [],
            search: [
              {
                column: "exporter_search",
                value: ` ${exporterName} `,
                comparison: "like",
                data: { },
                _key: `exporter_search_like_ ${exporterName} `
              }
            ],
            query_limit: { offset: 0, max_num_rows: 50 },
            hostDisplayType: "ip"
          })
        }
      };
    }
    

    };
    
    //params to figure out which interfaces exist for a device
   
  

  reportJSON(scrutInfo, scrutParams, filterObject = null) {

    
    
    let selectedGranularity = scrutParams.selectedGranularity
 
    if (filterObject){
      if (filterObject.granularity !== null){
        selectedGranularity = filterObject.granularity
      }
      
    }
 

    //returning report params to be passed into request
    return {
      url:scrutInfo['url'],
      'method':'get',
      params:{
        rm: "report_api",
        action: "get",
        authToken: scrutInfo['authToken'],
        rpt_json: JSON.stringify({
          reportTypeLang: scrutParams.reportType,
          reportDirections: {
            selected: scrutParams.reportDirection
          },
          times: {
            dateRange: "Custom",
            start: `${scrutParams.startTime}`,
            end: `${scrutParams.endTime}`
          },
          orderBy: scrutParams.scrutDisplay["display"],
          filters: scrutParams.scrutFilters,
          dataGranularity: {
            selected: selectedGranularity
          },
          showOthers: 0
        }),
  
        data_requested: JSON.stringify({
          [scrutParams.reportDirection]: {
            graph: "all",
            table: {
              query_limit: {
                offset: 0,
                max_num_rows: 10
              }
            }
          }
        })
      }

    };
  }  

}
export class Handledata {
  //scrutinizer returns graph data opposite of how grafana wants it. So we flip it here.
  constructor() {
    this.rearrangeData = (arr, oldIndex, newIndex) => {
      while (oldIndex < 0) {
        old_index += arr.length;
      }
      while (newIndex < 0) {
        new_index += arr.length;
      }
      if (newIndex >= arr.length) {
        let k = newIndex - arr.length;

        while (k-- + 1) {
          arr.push(undefined);
        }
      }
      arr.splice(newIndex, 0, arr.splice(oldIndex, 1)[0]);
      return arr;
    };
  }

  formatData(scrutData, scrutParams, graphGranularity, options, query=null) {

    let granularityChosen = graphGranularity
    //check if DNS resolve is on. 


    let dnsResolve = options.reportDNS

    if(query !== null){
     
    
      if(query.resolveDNS !== null ){
      
        dnsResolve = query.resolveDNS
      }
    }

   
    let graphRes = parseInt(granularityChosen) / 60

    let displayValue;

  
    if (scrutParams.scrutDisplay["display"] === "custom_interfacepercent") {
      displayValue = "percent";
    } else {
      displayValue = "bits";
    }

    let reportDirection = scrutParams.reportDirection;
    //grafana wants time in millaseconds. so we multiple by 1000.
    //we also want to return data in bits, so we device by 8
    let datatoGraph = [];

    let graphingData = scrutData;
    let i,
      j = 0;
    let graphData = graphingData["report"]["graph"]["pie"][reportDirection];

    let tableData =
      graphingData["report"]["graph"]["timeseries"][reportDirection];

    //if user is selecting bits, we need to multiple by 8, we also need to use the interval time.
    if (displayValue === "bits") {
      for (i = 0; i < tableData.length; i++) {
        for (j = 0; j < tableData[i].length; j++) {
          tableData[i][j][0] = tableData[i][j][0] * 1000;
          tableData[i][j][1] = (tableData[i][j][1] * 8) / (graphRes * 60);
          this.rearrangeData(tableData[i][j], 0, 1);
        }
      }
    } else {
      //since interface reporting uses the total tables, we dont need to math it.
      for (i = 0; i < tableData.length; i++) {
        for (j = 0; j < tableData[i].length; j++) {
          tableData[i][j][0] = tableData[i][j][0] * 1000;
          tableData[i][j][1] = Math.round(tableData[i][j][1]);
          this.rearrangeData(tableData[i][j], 0, 1);
        }
      }
    }

    for (i = 0; i < graphData.length; i++) {

      let interfaceId;
      let interfaceDesc;

      if (scrutParams["reportType"] === "interfaces") {
        if (scrutParams["reportDirection"] === "inbound") {
          interfaceId = "Inbound Interface";
          interfaceDesc = "Inbound";
        } else {
          interfaceId = "Outbound Interface";
          interfaceDesc = "Outbound";
        }
        //scrutinizer returns a small amout of "other traffic" for interface reporting
        //this has to do with the relationship between totals and conversations.
        //we don't need this data, so we toss it out. It makes it do we can use SingleStat
        //and Guage visualizations for interfaces, which is nice.
       
        if (graphData[i]["label"] != "Other") {
          //check to make sure there is utilization data for interfaces.
          if(tableData[i]){
            datatoGraph.push({
              target:
                interfaceDesc + "--" + graphData[i]["tooltip"][1][interfaceId],
              datapoints: tableData[i]
            });
          }

          
        }
      } else {

        if(!dnsResolve) {
          datatoGraph.push({
            target: graphData[i]["label"],
            datapoints: tableData[i]
          });
        } else {
          datatoGraph.push({
            target: graphData[i]["label_dns"],
            datapoints: tableData[i]
          });
        }



      }
    }

    return datatoGraph
    
  }

  formatOthers(formatedData){


  }

}


export class AdhocHandler {


  createObject(query, filterTypes, filters) {




    let filterObject = {
      sourceIp: [],
      exporterDetails: [],
      exporters: [],
      ports: [],
      destIp: [],
      others:null,
      granularity:null,
      resolve:null
    };
    
    
    query.adhocFilters.forEach(filter => {
      if (!filterTypes.includes(filter["key"])) {filterObject.exporters.push(filter["key"]);} 
      else if (filter["key"] === "Select Granularity"){
          let granularyValue = this.granularityTransform(filter['value'])
          filterObject.granularity = granularyValue} 
      else if (filter["key"] === "Show Others"){
          let showOthers = this.othersTransform(filter['value'])
          filterObject.others = showOthers
        }
      else if (filter["key"] === "Resolve DNS"){
        let resolveDNS = this.othersTransform(filter['value'])
        filterObject.resolve = resolveDNS
      }
      else{
        filters.forEach(filterType => {
          if (filterType["text"] === filter["key"]) {
            let filterKey = filterType["value"];
            let filterValue = filter["value"];
            filterObject[filterKey].push(filterValue);
          }
         });
       }
    });


  return filterObject
  }

granularityTransform(granularity){

  switch (granularity) {
    case "Auto":
      return "auto"
    case "1 Minute":
      return "1"
    case "5 Minute":
      return "5"
    case "30 Minute":
      return "30"
    case "2 Hour":
      return "120"
    case "12 Hour":
      return "720"
  }
}


othersTransform(showValue) {
  switch (showValue) {
    case "No":
      return false
    case "Yes":
      return true
    }

  }


}
